File: windows/CodexIsland.Core/QuotaModels.cs
using System.Globalization;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace CodexIsland.Core;

public sealed record QuotaWindow(
    double UsedPercent, int? WindowDurationMins = null, double? ResetsAt = null)
{
    public double RemainingPercent => Math.Clamp(100 - UsedPercent, 0, 100);
    public string PercentText => RemainingPercent is > 0 and < 1 ? "<1%"
        : Math.Floor(RemainingPercent).ToString(CultureInfo.InvariantCulture) + "%";
    public string PeriodTitle => WindowDurationMins switch
    {
        10080 => "本周",
        > 0 and var n when n % 1440 == 0 => $"{n / 1440}天",
        > 0 and var n when n % 60 == 0 => $"{n / 60}小时",
        > 0 and var n => $"{n}分钟",
        _ => "当前周期"
    };

    public string ResetText(DateTimeOffset now)
    {
        if (ResetsAt is not { } reset) return "恢复时间暂未提供";
        var seconds = reset - now.ToUnixTimeSeconds();
        if (seconds <= 0) return "已到恢复时间，等待刷新";
        // Keep untrusted, implausibly large timestamps out of integer conversions.
        if (seconds > TimeSpan.FromDays(3650).TotalSeconds) return "恢复时间暂未提供";
        var minutes = Math.Max(1, (long)Math.Ceiling(seconds / 60));
        return minutes switch
        {
            >= 1440 => $"{minutes / 1440}天{minutes % 1440 / 60}小时后恢复",
            >= 60 => $"{minutes / 60}小时{minutes % 60}分钟后恢复",
            _ => $"{minutes}分钟后恢复"
        };
    }
}

public sealed record QuotaBucket(string Id, string Name, QuotaWindow? Primary, QuotaWindow? Secondary)
{
    public IReadOnlyList<QuotaWindow> Windows => new[] { Primary, Secondary }.OfType<QuotaWindow>().ToArray();
}

public sealed record QuotaSnapshot(IReadOnlyList<QuotaBucket> Buckets, DateTimeOffset FetchedAt, string? PlanName)
{
    public QuotaBucket MainBucket => Buckets.FirstOrDefault(b => b.Id == "codex") ?? Buckets[0];

    public static QuotaSnapshot Demo(DateTimeOffset now) => new([
        new("codex", "Codex", new(28, 300, now.AddHours(2.4).ToUnixTimeSeconds()),
            new(46, 10080, now.AddDays(2.7).ToUnixTimeSeconds())),
        new("additional", "额外模型额度 · 示例", new(82, 10080, now.AddDays(3).ToUnixTimeSeconds()), null)
    ], now, "演示");
}

public enum QuotaError { MissingCodex, Disconnected, TimedOut, InvalidResponse, NoQuota, Service }

public sealed class QuotaException(QuotaError kind, string? message = null) : Exception(message ?? kind switch
{
    QuotaError.MissingCodex => "未找到 Codex，请先在 Windows 安装并登录，或从托盘选择 Codex 程序。",
    QuotaError.Disconnected => "与 Codex 的连接已断开，稍后会重试。",
    QuotaError.TimedOut => "读取超时，请检查网络后刷新。",
    QuotaError.InvalidResponse => "暂时无法识别 Codex 返回的额度信息。",
    QuotaError.NoQuota => "账号暂未返回订阅额度，请确认已用 ChatGPT 账号登录 Windows 版 Codex。",
    _ => "暂时无法读取额度，请检查 Codex 登录状态和网络。"
})
{
    public QuotaError Kind { get; } = kind;

    // Raw server errors can contain account data. Only these messages reach the UI.
    public static QuotaException FromServer(string message)
    {
        var text = message.ToLowerInvariant();
        if (text.Contains("api key") || text.Contains("apikey") || text.Contains("unsupported auth"))
            return new(QuotaError.Service, "当前登录方式未提供订阅额度，请用 ChatGPT 账号登录 Codex。");
        if (text.Contains("401") || text.Contains("auth") || text.Contains("login") || text.Contains("sign in"))
            return new(QuotaError.Service, "请先在 Windows 版 Codex 中登录 ChatGPT 账号，再点刷新。");
        if (text.Contains("429")) return new(QuotaError.Service, "查询过于频繁，稍后会自动重试。");
        return new(QuotaError.Service);
    }
}

public static class QuotaParser
{
    private static readonly JsonSerializerOptions Options = new() { PropertyNamingPolicy = JsonNamingPolicy.CamelCase };
    private sealed record WindowData([property: JsonRequired] double UsedPercent, int? WindowDurationMins, double? ResetsAt);
    private sealed record BucketData(string? LimitId, string? LimitName, WindowData? Primary, WindowData? Secondary, string? PlanType);
    private sealed record Response(BucketData? RateLimits, Dictionary<string, BucketData?>? RateLimitsByLimitId);

    public static QuotaSnapshot Parse(string json, DateTimeOffset now)
    {
        Response response;
        try { response = JsonSerializer.Deserialize<Response>(json, Options) ?? throw new JsonException(); }
        catch (JsonException) { throw new QuotaException(QuotaError.InvalidResponse); }
        var all = response.RateLimitsByLimitId ?? [];
        if (response.RateLimits is { } legacy) all.TryAdd(legacy.LimitId ?? "codex", legacy);
        var keys = all.Keys.OrderBy(k => k == "codex" ? 0 : 1).ThenBy(k => k, StringComparer.Ordinal).ToArray();
        var buckets = new List<QuotaBucket>();
        foreach (var key in keys)
        {
            if (all[key] is not { } bucket || (bucket.Primary is null && bucket.Secondary is null)) continue;
            buckets.Add(new(key, bucket.LimitName ?? (key == "codex" ? "Codex" : key),
                Convert(bucket.Primary), Convert(bucket.Secondary)));
        }
        if (buckets.Count == 0) throw new QuotaException(QuotaError.NoQuota);
        var plan = keys.Select(k => all[k]?.PlanType).FirstOrDefault(p => p is not null);
        return new(buckets, now, PlanLabel(plan));
    }

    private static QuotaWindow? Convert(WindowData? data)
    {
        if (data is null) return null;
        if (!double.IsFinite(data.UsedPercent) || (data.ResetsAt is { } reset && !double.IsFinite(reset)))
            throw new QuotaException(QuotaError.InvalidResponse);
        return new(data.UsedPercent, data.WindowDurationMins, data.ResetsAt);
    }

    private static string? PlanLabel(string? plan) => plan?.ToLowerInvariant() switch
    {
        null => null, "plus" => "Plus", "pro" => "Pro", "free" => "免费版",
        "team" or "business" => "团队版", "enterprise" => "企业版", "edu" => "教育版", _ => "ChatGPT"
    };
}

File: Sources/CodexIsland/QuotaModels.swift
import Foundation

struct QuotaWindow: Codable, Equatable {
    var usedPercent: Double
    var windowDurationMins: Int?
    var resetsAt: Double?

    var remainingPercent: Double { max(0, min(100, 100 - usedPercent)) }
    var periodTitle: String {
        guard let minutes = windowDurationMins, minutes > 0 else { return "当前周期" }
        if minutes == 10_080 { return "本周" }
        if minutes % 1440 == 0 { return "\(minutes / 1440)天" }
        if minutes % 60 == 0 { return "\(minutes / 60)小时" }
        return "\(minutes)分钟"
    }
    func resetText(now: Date) -> String {
        guard let reset = resetsAt else { return "恢复时间暂未提供" }
        let remaining = reset - now.timeIntervalSince1970
        guard remaining > 0 else { return "已到恢复时间，等待刷新" }
        let minutes = max(1, Int(ceil(remaining / 60)))
        if minutes >= 1440 { return "\(minutes / 1440)天\((minutes % 1440) / 60)小时后恢复" }
        if minutes >= 60 { return "\(minutes / 60)小时\(minutes % 60)分钟后恢复" }
        return "\(minutes)分钟后恢复"
    }
}

struct QuotaBucket: Identifiable, Equatable {
    var id: String
    var name: String
    var primary: QuotaWindow?
    var secondary: QuotaWindow?
}

struct QuotaSnapshot: Equatable {
    var buckets: [QuotaBucket]
    var fetchedAt: Date
    var planName: String?
    var mainBucket: QuotaBucket? {
        buckets.first(where: { $0.id == "codex" }) ?? buckets.first
    }
    static func demo(now: Date = Date()) -> QuotaSnapshot {
        QuotaSnapshot(buckets: [QuotaBucket(id: "codex", name: "Codex", primary:
            QuotaWindow(usedPercent: 28, windowDurationMins: 300, resetsAt: now.addingTimeInterval(8640).timeIntervalSince1970),
            secondary: QuotaWindow(usedPercent: 46, windowDurationMins: 10_080, resetsAt: now.addingTimeInterval(230400).timeIntervalSince1970))],
            fetchedAt: now, planName: "演示")
    }
}

enum QuotaFailure: Error, LocalizedError {
    case missingCodex, disconnected, timedOut, invalidResponse, noQuota, service(String)
    var errorDescription: String? {
        switch self {
        case .missingCodex: return "未找到 Codex，请先安装并登录 Codex。"
        case .disconnected: return "与 Codex 的连接已断开，稍后会重试。"
        case .timedOut: return "读取超时，请检查网络后刷新。"
        case .invalidResponse: return "暂时无法识别 Codex 返回的额度信息。"
        case .noQuota: return "账号暂未返回订阅额度，请确认已用 ChatGPT 账号登录 Codex。"
        case .service(let message): return message
        }
    }
    // Only controlled messages reach the UI; upstream errors may contain account data.
    static func fromServer(_ message: String) -> QuotaFailure {
        let text = message.lowercased()
        if text.contains("api key") || text.contains("apikey") || text.contains("unsupported auth") {
            return .service("当前登录方式未提供订阅额度，请用 ChatGPT 账号登录 Codex。")
        }
        if text.contains("401") || text.contains("auth") || text.contains("login") || text.contains("sign in") {
            return .service("请先在 Codex 中登录 ChatGPT 账号，再点刷新。")
        }
        if text.contains("429") { return .service("查询过于频繁，稍后会自动重试。") }
        return .service("暂时无法读取额度，请检查 Codex 登录状态和网络。")
    }
}

enum QuotaParser {
    private struct Response: Decodable {
        var rateLimits: Bucket?
        var rateLimitsByLimitId: [String: Bucket]?
    }
    private struct Bucket: Decodable {
        var limitId: String?
        var limitName: String?
        var primary: QuotaWindow?
        var secondary: QuotaWindow?
        var planType: String?
    }
    static func parse(_ data: Data, now: Date = Date()) throws -> QuotaSnapshot {
        let response: Response
        do { response = try JSONDecoder().decode(Response.self, from: data) }
        catch { throw QuotaFailure.invalidResponse }
        var all = response.rateLimitsByLimitId ?? [:]
        if let legacy = response.rateLimits {
            let key = legacy.limitId ?? "codex"
            if all[key] == nil { all[key] = legacy }
        }
        let keys = all.keys.sorted { a, b in
            if a == "codex" { return b != "codex" }
            if b == "codex" { return false }
            return a < b
        }
        let buckets = keys.compactMap { key -> QuotaBucket? in
            guard let value = all[key], value.primary != nil || value.secondary != nil else { return nil }
            return QuotaBucket(id: key, name: value.limitName ?? (key == "codex" ? "Codex" : key),
                               primary: value.primary, secondary: value.secondary)
        }
        guard !buckets.isEmpty else { throw QuotaFailure.noQuota }
        let plan = all["codex"]?.planType ?? keys.compactMap { all[$0]?.planType }.first
        return QuotaSnapshot(buckets: buckets, fetchedAt: now, planName: plan.map(planLabel))
    }
    private static func planLabel(_ plan: String) -> String {
        switch plan.lowercased() {
        case "plus": return "Plus"
        case "pro": return "Pro"
        case "free": return "免费版"
        case "team", "business": return "团队版"
        case "enterprise": return "企业版"
        case "edu": return "教育版"
        default: return "ChatGPT"
        }
    }
}

struct JSONLineBuffer {
    private var pending = Data()
    mutating func append(_ data: Data) throws -> [Data] {
        pending.append(data)
        guard pending.count <= 4 * 1024 * 1024 else { throw QuotaFailure.invalidResponse }
        var lines: [Data] = []
        while let newline = pending.firstIndex(of: 10) {
            let line = Data(pending[..<newline])
            pending.removeSubrange(...newline)
            if !line.isEmpty { lines.append(line) }
        }
        return lines
    }
}
